openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -408,24 +415,30 @@ def _refresh_addresses(self):
         handle cases in which the IP has changed
         """
         self.connection.device.refresh_from_db()
-        self.connection.refresh_from_db()
+        self.connection = self.connection.get_working_connection(self.connection.device)
         self.addresses = self.connection.get_addresses()
 
+    def _log_reconnecting_error(self, attempt):
+        addresses = ', '.join(self.addresses)
+        self.log(
+            _(
+                'Trying to reconnect to device at {addresses} (attempt n.{attempt})...'.format(
+                    addresses=addresses, attempt=attempt
+                )
+            ),
+            save=False,
+        )
+
     def _write_checksum(self, checksum):
         for attempt in range(1, self.RECONNECT_MAX_RETRIES + 1):
-            self._refresh_addresses()
-            addresses = ', '.join(self.addresses)
-            self.log(
-                _(
-                    'Trying to reconnect to device at {addresses} (attempt n.{attempt})...'.format(
-                        addresses=addresses, attempt=attempt
-                    )
-                ),
-                save=False,
-            )
             try:
-                self.connect()
-            except (NoValidConnectionsError, socket.timeout, SSHException) as error:
```

Comment:
We no longer need to catch these exceptions here. These are handled by DeviceConnection model 

https://github.com/openwisp/openwisp-controller/blob/6adafe3d84763c9e942ccceb6f4c006157e50208/openwisp_controller/connection/base/models.py#L345-L357